buffer/src/main/java/io/netty/buffer/api/BufferAllocator.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.buffer.api.pool.PooledBufferAllocator;
+
+import java.nio.ByteOrder;
+import java.util.function.Supplier;
+
+/**
+ * Interface for allocating {@link Buffer}s.
+ */
+public interface BufferAllocator extends AutoCloseable {
+    /**
+     * Produces a {@link BufferAllocator} that allocates unpooled, on-heap buffers.
+     * On-heap buffers have a {@code byte[]} internally, and their {@linkplain Buffer#nativeAddress() native address}
+     * is zero.
+     * <p>
+     * The concrete {@link Buffer} implementation is chosen by {@link MemoryManager#instance()}.
+     *
+     * @return A non-pooling allocator of on-heap buffers
+     */
+    static BufferAllocator heap() {
```

Comment:
An API mistake we made on gRPC was putting the factory method for the type on the type itself.   It lead to code that (adjusted for this Buffer API) looked like:

```java
BufferAllocator allocator = new BufferAllocatorThatCannotPossiblyBeDirect();
//...
buf = allocator.direct().allocate(1)
```

People ignore warnings, and will accidentally call the static method because that's what IntelliJ autocompletes for them.   My first impression of such a method is "oh, this will get the direct allocating version of this allocator", even though it will return a different type.

Reply:
Static methods on _interfaces_ cannot be called as if they are instance methods, so your code example would fail to compile.

Reply:
Sigh, you are right.   Been burned too many times with abstract classes with statics.